.ci-builds/.azure-pipelines.yml

```diff
@@ -394,3 +394,60 @@ jobs:
       env:
         CC: clang
         CXX: clang++
+#-------------------------------------------------------------------------
+- job: ICU4C_Cygwin_GCC_x86_64_Release
+  displayName: 'C: Cygwin GCC x86_64 Release'
+  timeoutInMinutes: 90
+  pool:
+    vmImage: 'windows-2019'
+  variables:
+    ICU_CI_CACHE: c:\icu-ci-cache
+    CYG_URL: https://cygwin.com/setup-x86_64.exe
+    CYG_MIRROR: http://mirrors.kernel.org/sourceware/cygwin/
+    CYG_PACKAGES: automake,gcc-core,gcc-g++,make,pkg-config,perl,python3
+    CYG_ROOT: c:\cygwin-root
+    CYG_CACHE: '$(ICU_CI_CACHE)\cygwin64-v3'
+    CYG_CACHED_SETUP: '$(CYG_CACHE)\setup.exe'
+  steps:
+    # Use 'autocrlf input' since checked-in files may already be using CRLF.
+    - script: |
+        git config --global core.autocrlf input
+      displayName: 'Configure Git to checkout with Unix line endings (LF)'
+    - checkout: self
+      lfs: true
+      fetchDepth: 10
+    # Cache expires after 7 days of no activity.
+    - task: Cache@2
+      displayName: 'Restore Cygwin cache'
+      inputs:
+        key: cygwin
```

Comment:
Hmm... I was reading a bit more about how the Azure CI caching works and it seems that it doesn't quite work like the AppVeyor caching.

https://docs.microsoft.com/en-us/azure/devops/pipelines/release/caching?view=azure-devops#use-cache-task
https://www.appveyor.com/docs/build-cache/

The Azure caches are immutable, meaning that once they are created you can't update/change them at all.
Whereas the AppVeyor caches are more like a "folder" that simply gets saved/restored each time the build is run.

The content of the Azure cache is tied to the "key". Since we use the same "key" to identify the cache each time, I think that this means we will forever be restoring the same (possibly old) cached data repeatedly. It also won't age out either, since there will likely be a build occurring within 7 days, causing a cache hit.

It looks like the general usage pattern for the Azure caches is to have the "key" value generated from the hashed contents of a ".lock" file (which lists out various packages and version numbers). However, I don't think there is really an equivalent for the Cygwin setup off-hand.

I'm trying to think of other ways that we could generate and/or use a new "key" value so that we won't end up with the same stale cache being used over and over again.

However, the other tricky part is that the file needs to exist when the `Cache@2` task first runs -- meaning that it effectively needs to be checked in to the git repo, as the Cygwin setup won't exist at all during the very first initial run...

I wonder if we could use the `.azure-pipelines.yml` file itself? The key would be the hashed content of the file, meaning that if anyone touches the file, then the cache would be updated. This would at least be better than it *never* being updated. Additionally, if we ever changed the set of `CYG_PACKAGES` (automake,gcc-core,gcc-g++,make,pkg-config,perl,python3) then I think we'd want to update the cache as well.

I think we could do this by changing the key to this:

```yml
    # Use the contents of the file ".azure-pipelines.yml" as part of the key, as that contains the list of CYG_PACKAGES.
    - task: Cache@2
      displayName: 'Restore Cygwin cache'
      inputs:
        key: 'cygwin | .ci-builds/.azure-pipelines.yml'
```

What do you think? 😊

Reply:
Thanks Jeff, you bring up a great point about the cache being immutable. I like your idea of using the `.azure-pipelines.yml` file itself so that we can update Cygwin if we changed `CYG_PACKAGES`. I was thinking in addition to that, we should include the Cygwin version as part of the key as well, so that when a new version is released we would be able to update the version here:

```yml
    - task: Cache@2
      displayName: 'Restore Cygwin cache'
      inputs:
        # Use the contents of the file ".azure-pipelines.yml" as part of the key, as that contains the list of CYG_PACKAGES.
        # Also include the Cygwin version as part of the key. If we want to use a newer version of Cygwin, we can update the key here.
        # Note: The version in the key may become out of sync with (older than) the version we actually use if we update this file
        #       without updating the key. Any updates to this file guarantees that we're use the latest version.
        key: 'cygwin-3.2.0 | .ci-builds/.azure-pipelines.yml'
```
This would effectively be the same since we need to update this file to change the version anyway, but provides a place to do so. How does this sound to you?